### ISSUE TITLE
[CAS 4.1.3] Exception happen during load test under JpaTicketRegistry mode 

### ISSUE THREAD

**Issue by richardwei2008** [Jasig/cas#1397]
![](http://images2015.cnblogs.com/blog/613455/201512/613455-20151230175330010-1175643337.png)

Here is the stack trace:

```
30-Dec-2015 17:33:11.385 SEVERE [http-nio-8443-exec-6] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet [cas] in context with path [/cas] threw exception [Request processing failed; nested exception is org.springframework.orm.jpa.JpaSystemException: Error while committing the transaction; nested exception is javax.persistence.RollbackException: Error while committing the transaction] with root cause
 org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : org.jasig.cas.ticket.ServiceTicketImpl.ticketGrantingTicket -> org.jasig.cas.ticket.TicketGrantingTicketImpl
	at org.hibernate.engine.spi.CascadingActions$8.noCascade(CascadingActions.java:398)
	at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:129)
	at org.hibernate.event.internal.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:167)
	at org.hibernate.event.internal.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:158)
	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:91)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:55)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1258)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:425)
	at org.hibernate.engine.transaction.internal.jdbc.JdbcTransaction.beforeTransactionCommit(JdbcTransaction.java:101)
	at org.hibernate.engine.transaction.spi.AbstractTransactionImpl.commit(AbstractTransactionImpl.java:177)
	at org.hibernate.jpa.internal.TransactionImpl.commit(TransactionImpl.java:77)
	at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerSynchronization.afterCommit(ExtendedEntityManagerCreator.java:457)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.invokeAfterCommit(TransactionSynchronizationUtils.java:133)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.triggerAfterCommit(TransactionSynchronizationUtils.java:121)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.triggerAfterCommit(AbstractPlatformTransactionManager.java:954)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:799)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:726)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:521)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:291)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy70.validateServiceTicket(Unknown Source)
	at org.jasig.cas.web.ServiceValidateController.handleRequestInternal(ServiceValidateController.java:169)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:146)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:50)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:967)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:858)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:843)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.jasig.cas.security.RequestParameterPolicyEnforcementFilter.doFilter(RequestParameterPolicyEnforcementFilter.java:250)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:344)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:261)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.jasig.inspektr.common.web.ClientInfoThreadLocalFilter.doFilter(ClientInfoThreadLocalFilter.java:62)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:85)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:344)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:261)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1096)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by mmoayyed** [comment]
@frett Have u run into something like this?

**Comment by frett** [comment]
nope, I haven't seen this exception before.

**Comment by mmoayyed** [comment]
Is this something you'd be willing and able to test, reproduce and debug?

**Comment by frett** [comment]
Yeah, I can look into it and see if I can find anything. Based on some initial research the exception is related to flushing nested objects. I'm curious if it's related to the extended persistence context that was enabled in 4.1.3, but will be disabled in 4.1.4. I'll get jmeter setup and do some testing locally to see if I can reproduce this.

**Comment by mmoayyed** [comment]
Thanks

**Comment by frett** [comment]
I spun up a jmeter login test against our current dev server and was unable to reproduce. I'm going to spin up a base config of 4.1.3 and test against that to see if I can reproduce there.

@richardwei2008 was this a single failure or did you see it multiple times? what load were you actually pushing through jmeter? was this close to stock CAS, or do you have a bunch of customizations in place?

**Comment by frett** [comment]
I was able to easily reproduce this with load on 4.1.3, but it is not happening on 4.1.4-SNAPSHOT. I'm fairly confident this was caused by the extended persistence context which has already been removed from 4.1.4

**Comment by mmoayyed** [comment]
Perfect. Thanks. 4.1.4 will be available in 2 weeks.

**Comment by richardwei2008** [comment]
Thanks all. @frett as long as you've reproduced it on 4.1.3, I am still going to answer your question:
-> was this a single failure or did you see it multiple times?
It's not a single failure but it happen multiple times. Moreover, it may happen continually and occasionally during one batch load. 
-> what load were you actually pushing through jmeter?
Actually I did not use jmeter but simple sahi from UI with the load of 20 threads incrementby 5 seconds. 
-> was this close to stock CAS, or do you have a bunch of customizations in place?
It's close to the repository. I don't have many customizations just follow configuration guidance.

I may still try it on 4.1.4 and 4.2rc. Would you kindly share your reproduce experience on 4.1.3 to me?

**Comment by frett** [comment]
My first tests were against the main overlay I work on which has quite a few customizations in place but wasn't showing the issue. That overlay was using a recent 4.1.4-SNAPSHOT build.

Since that wasn't working I then switched to a simple JPA overlay I had from diagnosing some other JPA bugs. I configured it to use 4.1.3, tossed a load of 50 simultaneous threads at it and immediately saw exceptions being thrown. Switching the overlay to 4.1.4-SNAPSHOT caused all the exceptions to disappear.
